pkg/controllers/cluster/kubeconfig_controller.go

```diff
@@ -0,0 +1,308 @@
+// SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Greenhouse contributors
+// SPDX-License-Identifier: Apache-2.0
+
+package cluster
+
+import (
+	"context"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/clientcmd"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+
+	greenhouseapis "github.com/cloudoperators/greenhouse/pkg/apis"
+	"github.com/cloudoperators/greenhouse/pkg/apis/greenhouse/v1alpha1"
+	"github.com/cloudoperators/greenhouse/pkg/clientutil"
+
+	"github.com/rxwycdh/rxhash"
+)
+
+type KubeconfigReconciler struct {
+	client.Client
+}
+
+//+kubebuilder:rbac:groups=greenhouse.sap,resources=clusters,verbs=get;list;watch
+//+kubebuilder:rbac:groups=greenhouse.sap,resources=organizations,verbs=get;list;watch
+//+kubebuilder:rbac:groups="",resources=secrets,verbs=get;list;watch
+// +kubebuilder:rbac:groups=greenhouse.sap,resources=cluster-kubeconfigs,verbs=get;list;watch;create;update;patch
+
+// SetupWithManager sets up the controller with the Manager
+func (r *KubeconfigReconciler) SetupWithManager(name string, mgr ctrl.Manager) error {
+	r.Client = mgr.GetClient()
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		Watches(&v1alpha1.Cluster{}, handler.EnqueueRequestsFromMapFunc(sameNameResource)).
+		Watches(&corev1.Secret{}, handler.EnqueueRequestsFromMapFunc(clusterSecretToCluster)).
+		Watches(&v1alpha1.Organization{}, handler.EnqueueRequestsFromMapFunc(r.organizationToClusters)).
+		Watches(&corev1.Secret{}, handler.EnqueueRequestsFromMapFunc(r.organizationSecretToClusters)).
+		For(&v1alpha1.ClusterKubeconfig{}).
+		Complete(r)
+}
+
+func (r *KubeconfigReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	l := log.FromContext(ctx).WithValues("cluster", req.Name, "namespace", req.Namespace)
+
+	var cluster v1alpha1.Cluster
+	if err := r.Get(ctx, req.NamespacedName, &cluster); err != nil {
+		l.Error(err, "unable to fetch cluster")
+		return ctrl.Result{}, err
+	}
+
+	if cluster.GetDeletionTimestamp() != nil {
+		l.Info("skip reconcile, cluster is being deleted")
+		return ctrl.Result{}, nil
+	}
+
+	var kubeconfig v1alpha1.ClusterKubeconfig
+	updateRequired := false
+	failed := false
+
+	if err := r.Get(ctx, req.NamespacedName, &kubeconfig); err != nil {
```

Comment:
There is this client utility: `clientutil.CreateOrPatch(..)`: https://github.com/cloudoperators/greenhouse/blob/main/pkg/clientutil/patch.go#L32

This could combine the create ClusterKubeconfig if not exists then reueue and update.

The patch part will compare the resource on the server with the locally mutated one and only apply changes on diff.

Reply:
fixed by bb49301e63c7e340c92b8baf9a440d4e47bfc0bb